### PR TITLE
Fix transmutes between pointers in different address spaces (e.g. fn ptrs on AVR)

### PR DESCRIPTION
**Pull request by erikdesjardins**
Currently, this causes a verifier error (https://godbolt.org/z/YYohed4bj), since it uses `bitcast`, which can't convert between address spaces.

Uncovered due to https://github.com/rust-lang/rust/pull/105545#discussion_r1045269309

r? @bjorn3

### CODE REVIEW

File: src/test/codegen/avr/avr-func-addrspace.rs [hunk 1]
```diff
@@ -87,3 +91,21 @@ pub extern "C" fn test() {
...
     }
 }
+
+// Validate that we can codegen transmutes between data ptrs and fn ptrs.
+
+// CHECK: define{{.+}}{{void \(\) addrspace\(1\)\*|ptr addrspace\(1\)}} @transmute_data_ptr_to_fn({{\{\}\*|ptr}}{{.*}} %x)
+#[no_mangle]
+pub unsafe fn transmute_data_ptr_to_fn(x: *const ()) -> fn() {
+    // It doesn't matter precisely how this is codegenned (as a roundtrip through memory or an addrspacecast),
+    // as long as it doesn't cause a verifier error, e.g. by using `bitcast` (which was the case before).
```
Comment (erikdesjardins) [inline src/test/codegen/avr/avr-func-addrspace.rs:101]: This is codegenned as
```ll
%0 = addrspacecast ptr %x to ptr addrspace(1)
```

I can add a check line instead if you'd prefer

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 2]
```diff
@@ -1810,15 +1810,20 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
+                let dst_is_ptr = dst_scalar.primitive() == abi::Pointer;
+                if src_is_ptr == dst_is_ptr {
                     assert_eq!(src.layout.size, dst.layout.size);
 
                     // NOTE(eddyb) the `from_immediate` and `to_immediate_scalar`
                     // conversions allow handling `bool`s the same as `u8`s.
                     let src = bx.from_immediate(src.immediate());
-                    let src_as_dst = bx.bitcast(src, bx.backend_type(dst.layout));
+                    // LLVM also doesn't like `bitcast`s between pointers in different address spaces.
+                    let src_as_dst = if src_is_ptr {
```
[the remark this answers is not in the document]
Comment (erikdesjardins) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:1822]: It doesn't, but that's not possible here due to `src_is_ptr == dst_is_ptr` above. I can change this line to `src_is_ptr && dst_is_ptr` if that would be clearer

[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:1822]: Right

**Pull request merged**